File: app/src/main/java/com/agrogames/islandsofwar/engine/impl/weapon/Weapon.java
package com.agrogames.islandsofwar.engine.impl.weapon;

import android.os.Build;

import androidx.annotation.RequiresApi;

import com.agrogames.islandsofwar.common.M;
import com.agrogames.islandsofwar.engine.abs.another.AnotherAdder;
import com.agrogames.islandsofwar.engine.abs.common.Cell;
import com.agrogames.islandsofwar.engine.abs.common.Point;
import com.agrogames.islandsofwar.engine.abs.bullet.BulletAdder;
import com.agrogames.islandsofwar.engine.abs.map.MapObject;
import com.agrogames.islandsofwar.engine.abs.unit.IUnit;
import com.agrogames.islandsofwar.engine.abs.map.MapProvider;
import com.agrogames.islandsofwar.engine.abs.unit.IUnitAdder;
import com.agrogames.islandsofwar.engine.abs.bullet.IBullet;
import com.agrogames.islandsofwar.engine.abs.weapon.IWeapon;
import com.agrogames.islandsofwar.engine.impl.bullet.Bullet;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Stream;

public class Weapon implements IWeapon {
    private final Point relativeLocation;
    private final float reload;
    private final String texture;
    private final float longRange;
    private final float rotationSpeed;
    private final int damage;
    private final float speed;
    private final int flightHeight;
    private final int targetHeight;
    private final Point[] bulletsStarts;
    private final boolean bang;
    private final String bulletTexture;

    private float fromLastReload = 0;
    private float relativeRotation;
    private Float goalRotation;
    private IUnit owner;

    public Weapon(Point relativeLocation, float reload, String texture, float longRange,
                  float rotationSpeed, Point[] bulletsStarts, String bulletTexture, int damage, float speed, int flightHeight, int targetHeight, boolean bang) {
        this.relativeLocation = relativeLocation;
        this.reload = reload;
        this.texture = texture;
        this.longRange = longRange;
        this.rotationSpeed = rotationSpeed;
        this.bulletTexture = bulletTexture;
        this.damage = damage * bulletsStarts.length;
        this.speed = speed;
        this.flightHeight = flightHeight;
        this.targetHeight = targetHeight;
        this.bulletsStarts = bulletsStarts;
        this.bang = bang;
    }

    @Override
    public Point getLocation() {
        Point rr = relativeLocation.rotate(owner.getRotation());
        return new Point(owner.getLocation().x + rr.x, owner.getLocation().y + rr.y);
    }

    public Point[] getBulletStarts(){
        List<Point> result = new ArrayList<>();
        Point l = getLocation();
        float rotation = getRotation();
        for(Point s : bulletsStarts){
            Point rr = s.rotate(rotation);
            result.add(new Point(l.x + rr.x, l.y + rr.y));
        }
        return result.toArray(new Point[0]);
    }

    @Override
    public float getRotation() {
        float r = owner.getRotation() + relativeRotation;
        if(r > Math.PI * 2f){
            r-=Math.PI * 2f;
        }
        if(r < 0){
            r+=Math.PI * 2f;
        }
        return r;
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    public void update(MapProvider provider, BulletAdder bulletAdder, IUnitAdder unitAdder, AnotherAdder anotherAdder, float deltaTime) {
        Point location = getLocation();
        Stream<IUnit> enemies = Arrays.stream(provider.getEnemies());
        IUnit[] near = enemies.filter(this::isNear).filter(e -> isAvailable(e, provider.getAll()))
                .sorted(Comparator.comparingInt(e -> (int) getDist(e.getLocation(), location))).toArray(IUnit[]::new);

        for (IUnit enemy: near){
            setRotation(enemy);
            break;
        }

        rotate(deltaTime);

        fromLastReload += deltaTime;
        if(fromLastReload < reload) return;

        for (IUnit enemy: near){
            setRotation(enemy);
            shoot(bulletAdder, enemy);
            return;
        }
    }

    private void shoot(BulletAdder bulletAdder, IUnit enemy) {
        if (!M.nearlyEquals(getRotation(), goalRotation)) return;
        fromLastReload = 0;
        Point enemyLocation = enemy.getTerritory().length == 0
                ? enemy.getLocation()
                : new Point(enemy.getTerritory()[0]);

        for(Point start : this.getBulletStarts()) {
            IBullet bullet = new Bullet(this.bulletTexture, start, this.speed, this.damage,
                    this.longRange, this.flightHeight, this.targetHeight, this.owner, this.bang);
            bullet.setGoal(enemyLocation);
            bulletAdder.addBullet(bullet);
        }
    }

    private void setRotation(IUnit enemy) {
        Point location = getLocation();
        Point enemyLocation = enemy.getLocation();
        goalRotation = getAngle(enemyLocation, location);
        if(goalRotation > Math.PI * 2f){
            goalRotation -= (float) (Math.PI * 2f);
        }
        if(goalRotation < 0){
            goalRotation += (float)Math.PI * 2f;
        }
    }

    private void rotate(float deltaTime){
        if(goalRotation == null) return;
        float r = getRotation();
        if(r > Math.PI * 2f){
            r-=Math.PI * 2f;
        }
        if(r < 0){
            r+=Math.PI * 2f;
        }
        if(r == goalRotation) return;
        if(r - goalRotation > Math.PI){
            r -= Math.PI * 2f;
        } else if(goalRotation - r > Math.PI){
            r += Math.PI * 2f;
        }
        if(r < goalRotation){
            r += Math.PI * deltaTime * rotationSpeed;
            if(r > goalRotation){
                r = goalRotation;
            }
        }
        else{
            r -= Math.PI * deltaTime * rotationSpeed;
            if(r < goalRotation){
                r = goalRotation;
            }
        }
        relativeRotation = r - owner.getRotation();
    }

    private boolean isAvailable(IUnit enemy, MapObject[] all){
        if(enemy.getMinDamage() > damage || enemy.getHeight() != targetHeight) return false;
        Point location = getLocation();
        Point enemyLocation = enemy.getLocation();
        float r = getAngle(enemyLocation, location);
        if(r > Math.PI * 2f){
            r-=Math.PI * 2f;
        }
        if(r < 0){
            r+=Math.PI * 2f;
        }
        float d = getDist(enemyLocation, location);

        float l = 0;
        float x = location.x;
        float y = location.y;

        while (l < d){
            l += 0.5f;
            x += Math.cos(r) * 0.5;
            y += Math.sin(r) * 0.5;

            for (MapObject mapObject: all) {
                if(mapObject != owner && mapObject != enemy && mapObject.getHeight() == flightHeight){
                    for (Cell cell: mapObject.getTerritory()){
                        Cell c = new Cell((int) x + 1, (int) y + 1);
                        if(cell.equals(c)) {
                            return false;
                        }
                    }
                }
            }
        }

        return true;
    }

    private float getAngle(Point p1, Point p2){
        float r = (float) Math.atan(
                ((double) p1.y - (double) p2.y) /
                        ((double) p1.x - (double) p2.x));
        if (p1.x  < p2.x){
            r += Math.PI;
        }
        return r;
    }

    private boolean isNear(IUnit enemy){
        Point el = enemy.getLocation();
        Point l = getLocation();
        float dist = getDist(el, l);
        return dist <= longRange;
    }

    private float getDist(Point p1, Point p2) {
        float dx = p1.x - p2.x;
        float dy = p1.y - p2.y;
        return (float) Math.sqrt(dx * dx + dy * dy);
    }

    @Override
    public String getTexture() {
        return texture;
    }

    @Override
    public void setOwner(IUnit owner) {
        this.owner = owner;
    }

    @Override
    public Point getRelativeLocation() {
        return relativeLocation;
    }

    @Override
    public float getLongRange() {
        return longRange;
    }

    @Override
    public int getDamage() {
        return damage;
    }
}

File: app/src/main/java/com/agrogames/islandsofwar/engine/impl/weapon/Weapon.kt
package com.agrogames.islandsofwar.engine.impl.weapon

import com.agrogames.islandsofwar.common.M
import com.agrogames.islandsofwar.engine.abs.bullet.BulletAdder
import com.agrogames.islandsofwar.engine.abs.bullet.IBullet
import com.agrogames.islandsofwar.engine.abs.common.Cell
import com.agrogames.islandsofwar.engine.abs.common.Point
import com.agrogames.islandsofwar.engine.abs.graphics.GraphicsAdder
import com.agrogames.islandsofwar.engine.abs.map.MapObject
import com.agrogames.islandsofwar.engine.abs.map.MapProvider
import com.agrogames.islandsofwar.engine.abs.unit.IUnit
import com.agrogames.islandsofwar.engine.abs.unit.IUnitAdder
import com.agrogames.islandsofwar.engine.abs.weapon.IWeapon
import com.agrogames.islandsofwar.engine.impl.bullet.Bullet

class Weapon(override val relativeLocation: Point, private val reload: Float, override val texture: String, override val longRange: Float,
             private val rotationSpeed: Float, private val bulletsStarts: Array<Point?>?, private val bulletTexture: String, damage: Int, private val speed: Float, private val flightHeight: Int, private val targetHeight: Int, private val bang: Boolean, private val bangRange: Float, private val bulletCreationSound: String) : IWeapon {
    override val damage: Int
    private var fromLastReload = 0f
    private var relativeRotation = 0f
    private var goalRotation: Float? = null
    private lateinit var owner: IUnit

    init {
        this.damage = damage * bulletsStarts!!.size
    }

    override val location: Point
        get() {
            val rr = relativeLocation.rotate(owner.rotation)
            return Point(owner.location.x + rr.x, owner.location.y + rr.y)
        }
    override val bulletStarts: Array<Point>
        get() {
            val result: MutableList<Point> = ArrayList()
            val l = location
            val rotation = rotation
            for (s in bulletsStarts!!) {
                val rr = s!!.rotate(rotation)
                result.add(Point(l.x + rr.x, l.y + rr.y))
            }
            return result.toTypedArray()
        }
    override val rotation: Float
        get() {
            var r = owner.rotation + relativeRotation
            if (r > Math.PI * 2f) {
                r -= (Math.PI * 2f).toFloat()
            }
            if (r < 0) {
                r += (Math.PI * 2f).toFloat()
            }
            return r
        }

    override fun update(provider: MapProvider, bulletAdder: BulletAdder, unitAdder: IUnitAdder, graphicsAdder: GraphicsAdder, deltaTime: Float) {
        val near = provider.enemies.filter { isNear(it) && isAvailable(it, provider.all) }
                .sortedBy {getDist(it.location, location).toInt() }
        for (enemy in near) {
            setRotation(enemy)
            break
        }
        rotate(deltaTime)
        fromLastReload += deltaTime
        if (fromLastReload < reload) return
        for (enemy in near) {
            setRotation(enemy)
            shoot(bulletAdder, enemy)
            return
        }
    }

    private fun shoot(bulletAdder: BulletAdder, enemy: IUnit) {
        if (!M.nearlyEquals(rotation, goalRotation!!)) return
        fromLastReload = 0f
        val enemyLocation = if (enemy.territory.size == 0) enemy.location else Point(enemy.territory[0])
        for (start in bulletStarts) {
            val bullet: IBullet = Bullet(bulletTexture, bulletCreationSound, start, speed, damage, longRange,
                    flightHeight, targetHeight, owner, bang, bangRange)
            bullet.setGoal(enemyLocation)
            bulletAdder.addBullet(bullet)
        }
    }

    private fun setRotation(enemy: IUnit) {
        val location = location
        val enemyLocation = enemy.location
        goalRotation = getAngle(enemyLocation, location)
        if (goalRotation!! > Math.PI * 2f) {
            goalRotation = goalRotation!! - (Math.PI * 2f).toFloat()
        }
        if (goalRotation!! < 0) {
            goalRotation = goalRotation!! + Math.PI.toFloat() * 2f
        }
    }

    private fun rotate(deltaTime: Float) {
        if (goalRotation == null) return
        var r = rotation
        if (r > Math.PI * 2f) {
            r -= (Math.PI * 2f).toFloat()
        }
        if (r < 0) {
            r += (Math.PI * 2f).toFloat()
        }
        if (r == goalRotation) return
        if (r - goalRotation!! > Math.PI) {
            r -= (Math.PI * 2f).toFloat()
        } else if (goalRotation!! - r > Math.PI) {
            r += (Math.PI * 2f).toFloat()
        }
        if (r < goalRotation!!) {
            r += (Math.PI * deltaTime * rotationSpeed).toFloat()
            if (r > goalRotation!!) {
                r = goalRotation!!
            }
        } else {
            r -= (Math.PI * deltaTime * rotationSpeed).toFloat()
            if (r < goalRotation!!) {
                r = goalRotation!!
            }
        }
        relativeRotation = r - owner.rotation
    }

    private fun isAvailable(enemy: IUnit, all: Array<MapObject>): Boolean {
        if (enemy.minDamage > damage || enemy.height != targetHeight) return false
        val location = location
        val enemyLocation = enemy.location
        var r = getAngle(enemyLocation, location)
        if (r > Math.PI * 2f) {
            r -= (Math.PI * 2f).toFloat()
        }
        if (r < 0) {
            r += (Math.PI * 2f).toFloat()
        }
        val d = getDist(enemyLocation, location)
        var l = 0f
        var x = location.x
        var y = location.y
        while (l < d) {
            l += 0.5f
            x += (Math.cos(r.toDouble()) * 0.5).toFloat()
            y += (Math.sin(r.toDouble()) * 0.5).toFloat()
            for (mapObject in all) {
                if (mapObject !== owner && mapObject !== enemy && mapObject.height == flightHeight) {
                    for (cell in mapObject.territory) {
                        val c = Cell(x.toInt() + 1, y.toInt() + 1)
                        if (cell == c) {
                            return false
                        }
                    }
                }
            }
        }
        return true
    }

    private fun getAngle(p1: Point?, p2: Point?): Float {
        var r = Math.atan(
                (p1!!.y.toDouble() - p2!!.y.toDouble()) /
                        (p1.x.toDouble() - p2.x.toDouble())).toFloat()
        if (p1.x < p2.x) {
            r += Math.PI.toFloat()
        }
        return r
    }

    private fun isNear(enemy: IUnit): Boolean {
        val el = enemy.location
        val l = location
        val dist = getDist(el, l)
        return dist <= longRange
    }

    private fun getDist(p1: Point?, p2: Point?): Float {
        val dx = p1!!.x - p2!!.x
        val dy = p1.y - p2.y
        return Math.sqrt((dx * dx + dy * dy).toDouble()).toFloat()
    }

    override fun setOwner(owner: IUnit) {
        this.owner = owner
    }
}